runtime/src/main/kotlin/kotlin/text/Strings.kt

```diff
@@ -228,7 +230,84 @@ public actual fun String(chars: CharArray): String = fromCharArray(chars, 0, cha
  * @throws IndexOutOfBoundsException if either [offset] or [length] are less than zero
  * or `offset + length` is out of [chars] array bounds.
  */
-public actual fun String(chars: CharArray, offset: Int, length: Int): String = fromCharArray(chars, offset, length)
+public actual fun String(chars: CharArray, offset: Int, length: Int): String {
+    if (offset < 0 || length < 0 || offset + length > chars.size)
+        throw ArrayIndexOutOfBoundsException()
+
+    return fromCharArray(chars, offset, length)
+}
+
+/**
+ * Concatenates characters in this [CharArray] or its subrange into a String.
+ *
+ * @param startIndex the beginning (inclusive) of the subrange of characters, 0 by default.
+ * @param endIndex the end (exclusive) of the subrange of characters, size of this array by default.
+ *
+ * @throws IndexOutOfBoundsException if [startIndex] is less than zero or [endIndex] is greater than the size of this array.
+ * @throws IllegalArgumentException if [startIndex] is greater than [endIndex].
+ */
+@SinceKotlin("1.3")
+@ExperimentalStdlibApi
+public actual fun CharArray.concatToString(startIndex: Int, endIndex: Int): String {
+    AbstractList.checkBoundsIndexes(startIndex, endIndex, size)
+    return fromCharArray(this, startIndex, endIndex - startIndex)
+}
+
+/**
+ * Returns a [CharArray] containing characters of this string or its substring.
+ *
+ * @param startIndex the beginning (inclusive) of the substring, 0 by default.
+ * @param endIndex the end (exclusive) of the substring, length of this string by default.
+ *
+ * @throws IndexOutOfBoundsException if [startIndex] is less than zero or [endIndex] is greater than the length of this string.
+ * @throws IllegalArgumentException if [startIndex] is greater than [endIndex].
+ */
+@SinceKotlin("1.3")
+@ExperimentalStdlibApi
+public actual fun String.toCharArray(startIndex: Int, endIndex: Int): CharArray {
+    AbstractList.checkBoundsIndexes(startIndex, endIndex, length)
+    return toCharArray(this, startIndex, endIndex - startIndex)
+}
+
+/**
+ * Decodes a string from the bytes in UTF-8 encoding in this array or its subrange.
+ *
+ * @param startIndex the beginning (inclusive) of the subrange to decode, 0 by default.
+ * @param endIndex the end (exclusive) of the subrange to decode, size of this array by default.
+ * @param throwOnInvalidSequence specifies whether to throw an exception on malformed byte sequence or replace it by the replacement char `\uFFFD`.
+ *
+ * @throws IndexOutOfBoundsException if [startIndex] is less than zero or [endIndex] is greater than the size of this array.
+ * @throws IllegalArgumentException if [startIndex] is greater than [endIndex].
+ * @throws CharacterCodingException if the byte array contains malformed UTF-8 byte sequence and [throwOnInvalidSequence] is true.
+ */
+public actual fun ByteArray.decodeToString(startIndex: Int, endIndex: Int, throwOnInvalidSequence: Boolean): String {
```

Comment:
annotations are missing

Reply:
Why do you believe those `@ExperimentalStdlibApi` makes sense here?

Reply:
It's an actual part of common experimental API. 
When we're adding a common-only experimental function, we do not make a dedicated effort to remove experimentality in Kotlin/Native only. So marking it experimental would be consistent with the other common-only experimental functions.

On the other hand making these functions non-experimental would allow us to deprecate the former utf8 conversion API earlier, which might be a good thing.